include/envoy/network/io_handle.h

```diff
@@ -0,0 +1,35 @@
+#pragma once
+
+#include <memory>
+
+#include "envoy/common/pure.h"
+
+namespace Envoy {
+namespace Network {
+
+/**
+ * IoHandle: an abstract interface for all I/O operations
+ */
+class IoHandle {
+public:
+  IoHandle() {}
+
+  virtual ~IoHandle() {}
+
+  // TODO(sbelair2) remove fd() method
+  virtual int fd() const PURE;
+
+  /**
+   * @param the socket file descriptor to set in the handle. Assigns an fd from
+   * an external socket operation such as from libevent or the dispatcher after construction
+   *
+   * TODO(sbelair2):  To be removed when the IoSocketHandle derivative is integrated
+   * and the fd is fully abstracted from clients.
+   */
+  virtual void operator=(int fd) PURE;
+};
+typedef std::shared_ptr<IoHandle> IoHandlePtr;
```

Comment:
Hmm, looking at the code, there are *many* instances of shared pointers that do not have Shared in their name. @alyssawilk- you know the practice here, which would you prefer?

Reply:
Apologies, I'd overlooked that this was a shared pointer.
What's the motivation for using a shared pointer, rather than having clear ownership?  I don't think we need multiple owners for this particular class.

Reply:
Well, there was a bug originally when a unique_ptr was used for IoHandle. But beyond that, looking at the current diffs, there are many instances where the IoHandlePtr is passed as a parameter, as the fd was passed around. If we use unique_ptr for IoHandle, then do we expect that we will move the IoHandle ptr in these situations? Just wondering...

Reply:
I would hope we'd have one owner of the IoHandle and we'd pass around IoHandle& references where appropriate, as we do with most (though not all!) other Envoy classes.

Reply:
Ok, I'll make that change and see if there are issues.

Reply:
@alyssawilk, @danzh2010 In case you're wondering what the delay is, I am still working on trying to convert IoHandlePtr to a unique_ptr, in particular with IoHandleConstPtr.

Reply:
I'm curious what is the use case of IoHandleConstPtr? I can imagine we will pass around IoHandle reference in many places, but there isn't many places we need to pass around IoHandlePtr. 

Another question please: who is the owner of an IoHandle instance? My propose is SocketImpl. What do you think?

Reply:
@danzh2010 There is no use yet for the IoHandleConstPtr- I believe that @alyssawilk suggested that we have it. Right now SocketImpl might own the IoHandle, but what happens when we have IoQuicHandles, or IoVppHandles, who will own it then? Wasn't the ultimate point of this exercise to have more than just a socket interface to Envoy?

Right now there are many issues relating to IoHandle being a unique_ptr- what do folks feel about my putting it back and putting a TODO(danzh2010) in there?

Reply:
What are the issues? What are the cases in which you need multiple owners? I think unique_ptr is the way to go.

Reply:
I'm resolving several sorts of issues- many centered around the IoHandleConstPtr being a unique_ptr. I've commented that out to get at the other issues with IoHandlePtr being unique and having the copy constructor being implicitly deleted. I don't see cases yet where there are multiple owners, there's no argument on that point so far.

Reply:
I *think* you need copy constructor because you are passing around IoHandlerPtr. I suggest to pass around IoHandle& instead, i.e. SocketImple::ioHandle() returns IoHandle& instead of IoHandlePtr. In this way, not many places will even use IoHandlePtr.

Reply:
As to the owner of IoHandle instance, QUIC probably will use the same SocketImpl, so we have no problem with making it the owner. It's upto VPP's implementation whether to re-use SocketImpl or to  put ownship in other impl classes. Are you considering about re-using current SocketImpl but without ownership of IoHandle for VPP? If that's the case, it's still possible to move current SocketImpl code without IoHandle unique_ptr into "SocketImplBase" and re-use that class. But it won't need to be done till you really need it.